transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapChannel.java

```diff
@@ -0,0 +1,300 @@
+            allocHandle.edgeTriggered(isFlagSet(Native.EPOLLET));
+
+            ByteBufAllocator allocator = config.getAllocator();
+
+            // Tell the allocator handle we're about to start another read loop.
+            allocHandle.reset(config);
+
+            epollInBefore();
+
+            try {
+                Throwable exception = null;
+                ByteBuf packetData = null;
+                LinuxSocket tunTapFd = tunTapFd();
+
+                try {
+                    // Repeatedly read packets from the interface for as long as there are
+                    // packets to be read and the allocator handle tells us its okay...
+                    do {
+                        // Allocate an appropriately sized buffer in which to receive data.
+                        packetData = allocHandle.allocate(allocator);
+
+                        // Tell the allocator handle how much data we're attempting to read.
+                        allocHandle.attemptedBytesRead(packetData.writableBytes());
+
+                        int writerIndex = packetData.writerIndex();
+                        long readRes;
+
+                        // If the buffer has an accessible memory address, read packet data directly into the
+                        // buffer memory. Otherwise, get the internal NIO buffer associated with the ByteBuf
+                        // and read into that.
+                        if (packetData.hasMemoryAddress()) {
+                            readRes = tunTapFd.readTunTapPacketAddress(packetData.memoryAddress(), writerIndex,
+                                packetData.capacity());
+                        } else {
+                            ByteBuffer nioData = packetData.internalNioBuffer(writerIndex, packetData.writableBytes());
+                            readRes = tunTapFd.readTunTapPacket(nioData, nioData.position(), nioData.limit());
+                        }
+
+                        // Stop reading if no packet was received.
+                        if (readRes == 0) {
+                            allocHandle.lastBytesRead(-1);
+                            packetData.release();
+                            packetData = null;
+                            break;
+                        }
+
+                        // Extract the length of the packet and the protocol from the read result.
+                        int packetLen = (int) (readRes & 0xFFFFFFFF);
+                        int protocol = (int) ((readRes >> 32) & 0xFFFF);
+
+                        // Advance the writer index in the packet data buffer by the length of the packet.
+                        packetData.writerIndex(writerIndex + packetLen);
+
+                        // Allocate a TunTapPacket object containing the packet data and the protocol information.
+                        TunTapPacket packet = new TunTapPacket(protocol, packetData);
+                        rcvdPackets.add(packet);
```

Comment:
can we directly call `pipeline.fireChannelRead(new TunTapPacket(protocol, packetData));` here? Note that the `fire*` methods won't throw. This may also simplify the call to `fireExceptionCaught` to just be in the `catch` statement.